btc-withdrawal/src/main/kotlin/withdrawal/btc/transaction/TransactionHelper.kt

```diff
@@ -126,15 +105,97 @@ class TransactionHelper(
             }
     }
 
+    /**
+     * Collects previously sent transactions, that may be used as an input for newly created transaction.
+     * It may go into recursion if not enough money for fee was collected.
+     * @param availableAddresses - set of addresses which transactions will be available to spend
+     * @param wallet - current wallet. Used to fetch unspents
+     * @param amount - amount of SAT to spend
+     * @param confidenceLevel - minimum depth of transactions
+     * @param recursivelyCollectedUnspents - list of unspents collected from all recursion levels. It will be returned at the end on execution
+     * @return list full of unspent transactions
+     */
+    private tailrec fun collectUnspentsRec(
+        availableAddresses: Set<String>,
+        wallet: Wallet,
+        amount: Long,
+        confidenceLevel: Int,
+        recursivelyCollectedUnspents: MutableList<TransactionOutput>
+    ): List<TransactionOutput> {
+        //Only confirmed transactions must be used
+        val unspents =
+            ArrayList<TransactionOutput>(wallet.unspents.filter { unspent ->
+                unspent.parentTransactionDepthInBlocks >= confidenceLevel
+                        //Cannot use already used unspents
+                        && !usedOutputs.contains(unspent)
+                        //Cannot use unspents from another level of recursion
+                        && !recursivelyCollectedUnspents.contains(unspent)
+            })
+        if (unspents.isEmpty()) {
+            throw IllegalStateException("Out of unspents")
+        }
+        /*
+        Wallet stores unspents in a HashSet. Order of a HashSet depends on several factors: current array size and etc.
+        This may lead different notary nodes to pick different transactions.
+        This is why we order transactions manually, essentially reducing the probability of
```

Comment:
You are reducing the probability, but it will be still not zero?

Reply:
I can't say for sure. There are too many factors to take into account. For example, we may be connected to extremely misbehaving Bitcoin node with completely different blockchain. So, yes. It's not zero.

Reply:
The peergroup node is our, so I doubt that it will be misbehaving